doc/source/contributing_docstring.rst

```diff
@@ -0,0 +1,634 @@
+`*args` and `**kwargs`, but not `self`.
+
+The parameters are defined by their name, followed by a space, a colon, another
+space, and the type (or types). Note that the space between the name and the
+colon is important. Types are not defined for `*args` and `**kwargs`, but must
+be defined for all other parameters. After the parameter definition, it is 
+required to have a line with the parameter description, which is indented, and
+can have multiple lines. The description must start with a capital letter, and
+finish with a dot.
+
+**Good:**
+
+.. code-block:: python
+
+    class Series:
+        def plot(self, kind, **kwargs):
+            """Generate a plot.
+
+            Render the data in the Series as a matplotlib plot of the
+            specified kind.
+
+            Parameters
+            ----------
+            kind : str
+                Kind of matplotlib plot.
+            **kwargs
+                These parameters will be passed to the matplotlib plotting
+                function.
+            """
+            pass
+
+**Bad:**
+
+.. code-block:: python
+
+    class Series:
+        def plot(self, kind, **kwargs):
+            """Generate a plot.
+
+            Render the data in the Series as a matplotlib plot of the
+            specified kind.
+
+            Note the blank line between the parameters title and the first
+            parameter. Also, not that after the name of the parameter `kind`
+            and before the colo, a space is missing.
+
+            Also, note that the parameter descriptions do not start with a
+            capital letter, and do not finish with a dot.
+
+            Finally, the `**kwargs` parameter is missing.
+
+            Parameters
+            ----------
+
+            kind: str
+                kind of matplotlib plot
+            """
+            pass
+
+Parameter types
+^^^^^^^^^^^^^^^
+
+When specifying the parameter types, Python built-in data types can be used
+directly:
+
+- int
+- float
+- str
+
+For complex types, define the subtypes:
+
+- list of [int]
+- dict of {str : int}
+- tuple of (str, int, int)
+- set of {str}
+
+In case there are just a set of values allowed, list them in curly brackets
+and separated by commas (followed by a space):
+
+- {0, 10, 25}
+- {'simple', 'advanced'}
+
+If the type is defined in a Python module, the module must be specified:
+
+- datetime.date
+- datetime.datetime
+- decimal.Decimal
+
+If the type is in a package, the module must be also specified:
+
+- numpy.ndarray
+- scipy.sparse.coo_matrix
+
+If the type is a pandas type, also specify pandas:
+
+- pandas.Series
+- pandas.DataFrame
+
+If more than one type is accepted, separate them by commas, except the
+last two types, that need to be separated by the word 'or':
+
+- int or float
+- float, decimal.Decimal or None
+- str or list of str
+
+If None is one of the accepted values, it always needs to be the last in
+the list.
```

Comment:
Maybe we should discuss how we want to have the "notion" of optional:

- int or float, optional
- in, float or None
- int or float, default None

Reply:
I vote option 3. I'd give a second place nod to option 1, but `optional` and `default None` are essentially the same thing. It seems like an unnatural rule to enforce that `default 'foo'` is the rule when a keyword has a non-`None` default argument but `optional` is the route for `None`

Reply:
> It seems like an unnatural rule to enforce that default 'foo' is the rule when a keyword has a non-None default argument but optional is the route for None

Having a different way to describe it, can also signal that it actually is something different in practice. In (many, but not all) cases, a value of None really means that it is not specified and is optional (like `method=None` in `fillna`, because the default is to fill with a fixed value and not to use a forward or backward filling method), in constrast with other keywords that have a default value (like `skipna=True`. For users it 'feels' like optional because you typically don't need to specify it, but it is not optional)